pkg/mcs/resource_manager/client/limiter.go

```diff
@@ -0,0 +1,401 @@
+	}
+	r := lim.reserveN(now, n, waitLimit)
+	return &r
+}
+
+// SetupNotificationAt enables the notification at the given threshold.
+func (lim *Limiter) SetupNotificationAt(now time.Time, threshold float64) {
+	lim.advance(now)
+	lim.notifyThreshold = threshold
+}
+
+// notify tries to send a non-blocking notification on notifyCh and disables
+// further notifications (until the next Reconfigure or StartNotification).
+func (lim *Limiter) notify() {
+	if lim.isLowProcess {
+		return
+	}
+	lim.notifyThreshold = 0
+	lim.isLowProcess = true
+	select {
+	case lim.lowTokensNotifyChan <- struct{}{}:
+	default:
+	}
+}
+
+// maybeNotify checks if it's time to send the notification and if so, performs
+// the notification.
+func (lim *Limiter) maybeNotify() {
+	if lim.IsLowTokens() {
+		lim.notify()
+	}
+}
+
+// IsLowTokens returns whether the limiter is in low tokens
+func (lim *Limiter) IsLowTokens() bool {
+	if lim.isLowProcess || (lim.notifyThreshold > 0 && lim.tokens < lim.notifyThreshold) {
+		return true
+	}
+	return false
+}
+
+// RemoveTokens decreases the amount of tokens currently available.
+func (lim *Limiter) RemoveTokens(now time.Time, amount float64) {
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+	now, _, tokens := lim.advance(now)
+	lim.last = now
+	lim.tokens = tokens - amount
+	lim.maybeNotify()
+}
+
+type tokenBucketReconfigureArgs struct {
+	NewTokens float64
+
+	NewRate float64
+
+	NotifyThreshold float64
+}
+
+// Reconfigure modifies all setting for limiter
+func (lim *Limiter) Reconfigure(now time.Time, args tokenBucketReconfigureArgs) {
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+	log.Debug("[resource group controllor] before reconfigure", zap.Float64("NewTokens", lim.tokens), zap.Float64("NewRate", float64(lim.limit)), zap.Float64("NotifyThreshold", args.NotifyThreshold))
+	now, _, tokens := lim.advance(now)
+	lim.last = now
+	lim.tokens = tokens + args.NewTokens
+	lim.limit = Limit(args.NewRate)
+	lim.notifyThreshold = args.NotifyThreshold
+	lim.isLowProcess = false
+	lim.maybeNotify()
+	log.Debug("[resource group controllor] after reconfigure", zap.Float64("NewTokens", lim.tokens), zap.Float64("NewRate", float64(lim.limit)), zap.Float64("NotifyThreshold", args.NotifyThreshold))
+}
+
+// AvailableTokens decreases the amount of tokens currently available.
+func (lim *Limiter) AvailableTokens(now time.Time) float64 {
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+	_, _, tokens := lim.advance(now)
+	return tokens
+}
+
+// reserveN is a helper method for Reserve.
+// maxFutureReserve specifies the maximum reservation wait duration allowed.
+// reserveN returns Reservation, not *Reservation, to avoid allocation in AllowN and WaitN.
+func (lim *Limiter) reserveN(now time.Time, n float64, maxFutureReserve time.Duration) Reservation {
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+
+	if lim.limit == Inf {
+		return Reservation{
+			ok:        true,
+			lim:       lim,
+			tokens:    n,
+			timeToAct: now,
+		}
+	} else if n > maxRequestTokens {
```

Comment:
I think `maxRequestTokens` should be configuable , or no need maxRequestTokens  in client side.